In [1]:
import numpy as np

In [2]:
def combinations(n: int, k: int) -> int:
    return np.math.factorial(n) // (np.math.factorial(k) * np.math.factorial(n - k))

### Задача 1

#### Имеется колода в 52 карты. Найти число возможностей вытянуть из неё 4 карты так, чтобы среди них был хотя бы один туз.

In [3]:
ace_1 = combinations(4, 1) # 1 туз из 4 возможных
ace_2 = combinations(4, 2) # 2 туза из 4 возможных
ace_3 = combinations(4, 3) # 3 туза из 4 возможных
ace_4 = combinations(4, 4) # 4 туза из 4 возможных
not_ace_1 = combinations(48, 3) # 3 не туза из оставшихся 48 карт (всего карт 52)
not_ace_2 = combinations(48, 2) # 2 не туза из оставшихся 48 карт (всего карт 52)
not_ace_3 = combinations(48, 1) # 1 не туза из оставшихся 48 карт (всего карт 52)
not_ace_4 = combinations(48, 0) # 0 не тузов из оставшихся 48 карт (всего карт 52)

ace_1 = $C_4^1$, ace_2 = $C_4^2$, ace_3 = $C_4^3$, ace_4 = $C_4^4$

not_ace_1 = $C_{48}^3$, not_ace_1 = $C_{48}^2$, not_ace_1 = $C_{48}^1$, not_ace_1 = $C_{48}^0$

In [4]:
result = ace_1 * not_ace_1 + ace_2 * not_ace_2 + ace_3 * not_ace_3 + ace_4 * not_ace_4
print(result)

76145


result = $C_4^1 \cdot C_{48}^3 + C_4^2 \cdot C_{48}^2 + C_4^3 \cdot C_{48}^1 + C_4^4 \cdot C_{48}^0 = 76145$

Через цикл:

In [5]:
result = sum(combinations(4, i) * combinations(48, 4 - i) for i in range(1, 5))
print(result)

76145


result = $C_4^i \cdot C_{48}^{4 - i} = 76145$

Через вычитание числа исходов, при которых в руке нет тузов, от общего числа исходов:

In [6]:
result = combinations(52, 4) - combinations(48, 4)
print(result)

76145


result = $C_{52}^4 - C_{48}^4 = 76145$

### Задачи 2 и 3

#### Из 60 вопросов, входящих в экзаменационные билеты, студент знает 50. Случайным образом студент вытягивает 3 вопроса. Какова вероятность, что все выбранные вопросы знакомы студенту? Какова вероятность что два из трёх вопросов знакомы студенту?

In [7]:
all_combinations = combinations(60, 3)
true_combinations = combinations(50, 3)

all_combinations = $C_{60}^3$

true_combinations = $C_{50}^3$

In [8]:
result_1 = true_combinations / all_combinations # вероятность, что все 3 знакомы
print(result_1)

0.5727644652250146


result_1 = $\frac{C_{50}^3}{C_{60}^3} = 0.5727644652250146$

In [9]:
result_2 = combinations(50, 2) * combinations(10, 1) / all_combinations # вероятность, что 2 из 3 знакомы
print(result_2)

0.35797779076563413


result_2 = $\frac{C_{50}^2 \cdot C_{10}^1}{C_{60}^3} = 0.35797779076563413$

### Задача 4

#### Допустим, имеется некоторая очень редкая болезнь (поражает 0.1% населения). Вы приходите к врачу, вам делают тест на эту болезнь, и тест оказывается положительным. Врач говорит вам, что этот тест верно выявляет 99% больных этой болезнью и всего лишь в 1% случаев даёт ложный положительный ответ.

#### Вопрос: какова вероятность, что вы действительно больны ей?

#### Подсказка: используйте формулу Байеса с раскрытием знаменателя с помощью формулы полной вероятности.

In [10]:
P_a = 0.001
P_not_a = 0.999
P_b_a = 0.99
P_b_not_a = 0.01

$P(A|B) = \frac{P(B|A) \cdot P(A)}{P(B)}$

In [11]:
P_b = P_b_a * P_a + P_b_not_a * P_not_a
print(P_b)

0.01098


$P(B) = P(B|A) \cdot P(A) + P(B|\overline{A}) \cdot P(\overline{A}) = 0.01098$

In [12]:
P_a_b = P_b_a * P_a / P_b
print(P_a_b)

0.09016393442622951


$P(A|B) = \frac{P(B|A) \cdot P(A)}{P(B|A) \cdot P(A) + P(B|\overline{A}) \cdot P(\overline{A})} = 0.09016393442622951$